rdr_service/dao/questionnaire_response_dao.py

```diff
@@ -758,6 +777,55 @@ def _update_participant_summary(
                     session, participant.participantId, gender_code_ids
                 )
 
+    def _is_digital_health_share_code(self, code_value):
+        return code_value.lower() in [APPLE_EHR_SHARING_MODULE, APPLE_EHR_STOP_SHARING_MODULE,
+                                      APPLE_HEALTH_KIT_SHARING_MODULE, APPLE_HEALTH_KIT_STOP_SHARING_MODULE,
+                                      FITBIT_SHARING_MODULE, FITBIT_STOP_SHARING_MODULE]
+
+    def _update_digital_health_status_field(self, current_value, code_value, authored):
+        something_changed = False
+        authored_str = format_datetime(authored)
+        field_mapping = {
+            APPLE_HEALTH_KIT_SHARING_MODULE: ('appleHealthKit', 'YES'),
+            APPLE_HEALTH_KIT_STOP_SHARING_MODULE: ('appleHealthKit', 'NO'),
+            APPLE_EHR_SHARING_MODULE: ('appleEHR', 'YES'),
+            APPLE_EHR_STOP_SHARING_MODULE: ('appleEHR', 'NO'),
+            FITBIT_SHARING_MODULE: ('fitbit', 'YES'),
+            FITBIT_STOP_SHARING_MODULE: ('fitbit', 'NO')
+        }
+        current_value = current_value if current_value is not None else {}
+        # sqlalchemy can't update the Json field directly, need a deepcopy to replace the old value
+        new_value = copy.deepcopy(current_value)
+        if field_mapping[code_value][0] in new_value:
+            current_history = new_value[field_mapping[code_value][0]]['history']
+            exist = False
+            for item in current_history:
+                if item['status'] == field_mapping[code_value][1] and item['authoredTime'] == authored_str:
+                    exist = True
+                    break
+            if not exist:
+                new_value[field_mapping[code_value][0]]['status'] = field_mapping[code_value][1]
```

Comment:
Possible data issues: In the chance that PTSC sends us payloads out of order (like if the latest for a participant is a yes, but we get an older payload that was 'no', but just has a slightly different authored time) should we make sure that what we're setting as current isn't older than the current we already have?

Reply:
Good catch, done!